src/realm/realm.h

```diff
@@ -0,0 +1,1612 @@
+ *
+ * This is intended as a migration path for users of the C++ Object Store API.
+ *
+ * Call `realm_release()` on the returned `realm_t*` to decrement the refcount
+ * on the inner `std::shared_ptr<Realm>`.
+ *
+ * @param pshared_ptr A pointer to an instance of `std::shared_ptr<Realm>`.
+ * @param n Must be equal to `sizeof(std::shared_ptr<Realm>)`.
+ * @return A `realm_t*` representing the same Realm object as the passed
+ *         `std::shared_ptr<Realm>`.
+ */
+RLM_API realm_t* _realm_from_native_ptr(const void* pshared_ptr, size_t n);
+
+/**
+ * Forcibly close a Realm file.
+ *
+ * Note that this invalidates all Realm instances for the same path.
+ *
+ * The Realm will be automatically closed when the last reference is released,
+ * including references to objects within the Realm.
+ *
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_close(realm_t*);
+
+/**
+ * True if the Realm file is closed.
+ *
+ * This function cannot fail.
+ */
+RLM_API bool realm_is_closed(realm_t*);
+
+/**
+ * Begin a write transaction for the Realm file.
+ *
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_begin_write(realm_t*);
+
+/**
+ * Return true if the realm is in a write transaction.
+ *
+ * This function cannot fail.
+ */
+RLM_API bool realm_is_writable(const realm_t*);
+
+/**
+ * Commit a write transaction.
+ *
+ * @return True if the commit succeeded and no exceptions were thrown.
+ */
+RLM_API bool realm_commit(realm_t*);
+
+/**
+ * Roll back a write transaction.
+ *
+ * @return True if the rollback succeeded and no exceptions were thrown.
+ */
+RLM_API bool realm_rollback(realm_t*);
+
+/**
+ * Refresh the view of the realm file.
+ *
+ * If another process or thread has made changes to the realm file, this causes
+ * those changes to become visible in this realm instance.
+ *
+ * This calls `advance_read()` at the Core layer.
+ *
+ * @return True if the realm was successfully refreshed and no exceptions were
+ *         thrown.
+ */
+RLM_API bool realm_refresh(realm_t*);
+
+/**
+ * Produce a frozen view of this realm.
+ *
+ * @return A non-NULL realm instance representing the frozen state.
+ */
+RLM_API realm_t* realm_freeze(realm_t*);
+
+/**
+ * Vacuum the free space from the realm file, reducing its file size.
+ *
+ * @return True if compaction was successful and no exceptions were thrown.
+ */
+RLM_API bool realm_compact(realm_t*, bool* did_compact);
+
+/**
+ * Create a new schema from classes and their properties.
+ *
+ * Note: This function does not validate the schema.
+ *
+ * Note: `realm_table_key_t` and `realm_col_key_t` values inside
+ *       `realm_class_info_t` and `realm_property_info_t` are unused when
+ *       defining the schema. Call `realm_get_schema()` to obtain the values for
+ *       these fields in an open realm.
+ *
+ * @return True if allocation of the schema structure succeeded.
+ */
+RLM_API realm_schema_t* realm_schema_new(const realm_class_info_t* classes, size_t num_classes,
```

Comment:
Consider using `const realm_class_info_t classes[]` and `const realm_property_info_t* class_properties[]` to signal that we are expecting arrays here. This comment may applies everywhere we are expecting an array.

Reply:
I like it on principle, but I haven't seen it used in many popular C libraries, and I'm worried about confusing less C-literate users of the API. If you really like it, I would be fine with it, but I think I'll merge without it for now.